### PR TITLE
changing expander regex 

### PR DESCRIPTION
**Pull request by SimaTian**
to use code generated version instead of compiled one where applicable according to this[ .NET article.](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-source-generators)
This nets a small but visible performance gain with only a minor code update.

### CODE REVIEW

**Comment by JanKrivanek** [comment]
Contributes to https://github.com/dotnet/msbuild/issues/7598

File: src/Build/Evaluation/Expander.cs [hunk 1]
```diff
@@ -3086,24 +3086,40 @@ internal string GetMetadataValueFromMatch(Match match)
...
             /// </summary>
             internal static readonly Lazy<Regex> ItemMetadataPattern = new Lazy<Regex>(
                 () => new Regex(ItemMetadataSpecification,
                     RegexOptions.IgnorePatternWhitespace | RegexOptions.ExplicitCapture | RegexOptions.Compiled));
+#endif
 
-            /// <summary>
-            /// Name of the group matching the "name" of a metadatum.
-            /// </summary>
+            internal static Regex ItemMetadataPatternWrapper()
```
Comment (JanKrivanek) [inline src/Build/Evaluation/Expander.cs:3111]: Should this rather be a getter? (it's a bit weird to see a method name that doesn't have any verb)

**Review by JanKrivanek — APPROVED** [review]
Thank you!

If you have any measurements you run in your env - feel free to attach here and brag more

File: src/Build/Evaluation/Expander.cs [hunk 2]
```diff
@@ -3128,6 +3152,15 @@ private static class RegularExpressions
                     ItemVectorWithTransformLHS + @")" + ItemMetadataSpecification + @"(?!" +
                     ItemVectorWithTransformRHS + @"))",
                     RegexOptions.IgnorePatternWhitespace | RegexOptions.ExplicitCapture | RegexOptions.Compiled));
+#endif
+            internal static Regex NonTransformItemMetadataPatternWrapper()
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:3156]: it looks like a getter candidate too :)

File: src/Build/Evaluation/Expander.cs [hunk 3]
```diff
@@ -3116,6 +3132,14 @@ private static class RegularExpressions
             /// </summary>
             internal const string ItemTypeGroup = "ITEM_TYPE";
 
+#if NET7_0_OR_GREATER
+            [GeneratedRegex(@"((?<=" + ItemVectorWithTransformLHS + @")" + ItemMetadataSpecification + @"(?!" +
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:3136]: Is it possible to move this string to some const and reuse it in these 2 places?
```
@"((?<=" + ItemVectorWithTransformLHS + @")" + ItemMetadataSpecification + @"(?!" +
                            ItemVectorWithTransformRHS + @")) | ((?<!" + ItemVectorWithTransformLHS + @")" +
                            ItemMetadataSpecification + @"(?=" + ItemVectorWithTransformRHS + @")) | ((?<!" +
                            ItemVectorWithTransformLHS + @")" + ItemMetadataSpecification + @"(?!" +
                            ItemVectorWithTransformRHS + @"))"
```

**Comment by YuliiaKovalova** [comment]
It looks good!
I have found other places with Regex usage in MSBuild:
https://github.com/search?q=repo%3Adotnet%2Fmsbuild%20%20Lazy%3CRegex%3E&type=code

Do you plan to adapt them too to the new practice?

**Review by YuliiaKovalova — APPROVED** [review]

**Comment by SimaTian** [comment]
> It looks good! I have found other places with Regex usage in MSBuild: https://github.com/search?q=repo%3Adotnet%2Fmsbuild%20%20Lazy%3CRegex%3E&type=code
> 
> Do you plan to adapt them too to the new practice?

Thanks.
I will take a look. The new pattern is easily applicable only when the pattern is known at a compile time. E.g. constant or with a limited enough pool of options to be reasonably enumerated by hand so it might not be as straightforward.
I started with this one as it is used quite a bit within the context of the expander.

I noticed this one as it was visible in profiler, that being said, unifying the rest is a good idea - I will see what I can do.
before:
regex-before

after:
regex-after

so the impact in context of MSBUild is in the range of ~0.1%, but still visible. 
The speedup is much more visible when doing targeted testing:
```
static partial class Test
{
    internal const string itemTypeOrMetadataNameSpecification = @"[A-Za-z_][A-Za-z_0-9\-]*";

    // the portion of an item transform that is the function that we wish to execute on the item
    internal const string itemFunctionNameSpecification = @"[A-Za-z]*";

    private const string ItemMetadataSpecification = @"%\(\s* (?<ITEM_SPECIFICATION>(?<ITEM_TYPE>" + itemTypeOrMetadataNameSpecification + @")\s*\.\s*)? (?<NAME>" + itemTypeOrMetadataNameSpecification + @") \s*\)";
    
    [GeneratedRegex(ItemMetadataSpecification)]
    private static partial Regex GeneratedPattern();

    public static Regex ItemMetadataPattern = new Regex(ItemMetadataSpecification,
                   RegexOptions.IgnorePatternWhitespace | RegexOptions.ExplicitCapture | RegexOptions.Compiled );

    public static string benchmark_regex(string expression)
    {
        return Test.ItemMetadataPattern.Replace(expression, new MatchEvaluator(matchEvaluator.ExpandSingleMetadata));
        //return Test.GeneratedPattern().Replace(expression, new MatchEvaluator(matchEvaluator.ExpandSingleMetadata));
    }
}
```
old, compiled regex:
regex-compiled

new, generated regex:
regex-generated

Now my test strings for matching are not all that representative so this targeted illustration isn't pefect:
```
string[] tests = { "some random string", "%(test.test)", "%(test.test.test)%(test.test)%(test.test)" , "asdfa.sdfasdfa", "%dasfsd.2dsfa", "%(dasfsd2.dsfa)" };
var match = 0;
for (int i = 0; i != 10000000; i++)
{
    foreach (var test in tests)
    {
        //Console.WriteLine(Test.benchmark_regex(test));
        if (Test.benchmark_regex(test).StartsWith("y", StringComparison.Ordinal));
        {
            match++;
        }
    }
}
```
Namely in Expander, we're already doing some fail-soon pre-scanning so some of the edge cases might be slightly different.

I'm pondering how much more we could gain by tweaking the Regex.Replace itself, but that is in the realm of speculation.
